csrc/alias_analysis.cpp

```diff
@@ -429,15 +432,21 @@ bool okToRelayout(
 } // namespace
 
 void AliasAnalysisResult::finalize(
-    const bool can_override_empty_allocation_domain) {
+    const bool can_override_empty_allocation_domain,
+    const bool allow_output_alias_intermediate) {
   for (auto [alias, root_and_layout] : alias_to_source_) {
     auto [root, preferred_layout] = root_and_layout;
-    // Walks up the `alias_to_source_` chain.
-    while (root != nullptr && !root->isFusionInput() &&
-           !root->isFusionOutput()) {
-      const auto i = alias_to_source_.find(root);
-      root = (i == alias_to_source_.end() ? nullptr : i->second.first);
+    // If allow output to alias intermediate, then we don't need to walk up
+    // the chain to find an input or output root. It changes the last reshape
+    // in group norm to a no-op.
+    if (!allow_output_alias_intermediate) {
```

Comment:
Thanks for the patch. Have you read https://github.com/NVIDIA/Fuser/issues/2395#issuecomment-2207043749? I tagged you there but the notification apparently didn't land. Would that solve this groupnorm issue?

Reply:
Yes and it looks great. IIUC, step-1 in the proposal is `Run alias analysis as we do today`, however, current `AliasAnalysisResult::finalize` needs to walk up the chain to find an input or output root and it will miss the case in group norm where the output alias an intermediate tensor. That's why I need to add this patch.
Previously, I also need to patch `MarkAliasesPreparePass::runPass` (see [commit](7175220caf004e979a603deb0b78f5e16643bdce) ), but with #2529, that path is no longer required.

Reply:
> superseded



> > AliasAnalysisResult::finalize needs to walk up the chain to find an input or output root and it will miss the case in group norm where the output alias an intermediate tensor.
> 
> Yes, that's an important implementation detail that needs to be reworked. `alias_to_source_` does keep aliases between intermediate tensors. So I'll have to put `segment_set` based on that without having to find the fusion I/O.
> 
> Overall, I feel lots of this PR is going to be superseded. However, if you need it in a week, I don't mind reviewing it. Wdyt?

Thanks. It would be great to have this PR merged, so we can make group norm works as expected. Assuming it won't cause troubles for your planned works to revise alias analysis.